content/english/blog/caps_are_effects.md

```diff
@@ -0,0 +1,60 @@
+
+```
+internationalWithdrawal account amount localCurrency =
+    exchangeRate = Market.getRate(localCurrency)
+    newAmount = amount * exchangeRate
+    fee = Market.getFee(localCurrency)
+    Ledger.withdraw(account, amount, newAmount, exchangeRate)
+    Ledger.fee(account, fee)
+    Ledger.balance(account)
+```
+This function reaches out to "the market" to get the currency exchange rate between the canonical currency and the local currency. Then it reaches out to get the current fee for international withdrawals. Finally, it consumes a ledger to perform the withdrawal, take away the international transaction fee, and finally return the current/updated balance.
+
+There's nothing pure about this function, but it's an _extremely_ common idiom. This function relies on two external interactions, a `Market` and a `Ledger`. The exchange rate is something that fluctuates constantly throughout the day, and the ledger presumably gives the function access to the account ledger for a specific account.
+
+This function is non-deterministic because it has a number of _algebraic effects_. Such effects are a category for the messy, "impure" things that happen in our functions. For most of us, we probably write more impure functions than not.
+
+I've already tried to make this function fairly clean and simple. In contrast, I've seen functions like this turned into "kitchen sinks" where a single function establishes a connection to two different databases and issues queries directly to them. We fool ourselves into thinking the function is pure by hiding the tight coupling a few layers down in the library, but this is still tight coupling at its worst. In cases like this, you can't test this function without live access to real databases, and after that you have to figure out how to make your tests deterministic (which often involves "test databases").
+
+A lot of us are used to patterns like (micro)services, abstractions, anti-corruption layers, and more all designed to help us mitigate the ugly side effects this function has. But what if we could embrace these effects and write functions that are explicit about their effect needs, _without_ losing testability, flexibility, and purity?
+
+In a traditional object-oriented language or framework, we might treat each of these "effect providers" as an interface and then use something like dependency injection to shunt in an implementation for the effect at runtime (and presumably shunt in a mock during test time).
+
+In wasmCloud, we manage algebraic effects through [capability providers](https://wasmcloud.dev/reference/host-runtime/capabilities/). Here the capability provider, as seen by the WebAssembly module ([actor](https://wasmcloud.dev/reference/host-runtime/actors/)), is just an abstraction. It's a versioned contract through which the WebAssembly function gets its effects.
```

Comment:
If you surround the hyperlinks here with `<u>` tags they will actually show as underlined